### DISCUSSION TITLE
JSON Path

### DISCUSSION THREAD

**Discussion by ohler55** [ohler55/ojg#59] [General]
JSON Path is super useful for navigating not only JSON but data held in memory. Let's keep discussions of JSON Path here starting with this [comparison](https://cburgmer.github.io/json-path-comparison) of many implementations across a variety of languages.

**Comment by obloquy** [comment]
Is jsonpath filtering on a struct (as opposed to map) supported?
I didn't want to report this as an issue since I don't know what the intended effect should be. Adding the following to jp/get_test.go shows the difference.

```
diff --git a/jp/get_test.go b/jp/get_test.go
index 1c5cb57..1daa94d 100644
--- a/jp/get_test.go
+++ b/jp/get_test.go
@@ -184,6 +184,12 @@ var (
                {path: "$[-10:]", expect: []interface{}{1, 2, 3}, data: []int{1, 2, 3}},
                {path: "$[1:-10:-1]", expect: []interface{}{1, 2}, data: []int{1, 2, 3}},`
                {path: "$[2:10]", expect: []interface{}{3}, data: []int{1, 2, 3}},
+               // filter with map
+               {path: "$.x[?(@.b=='sample1')].a", expect: []interface{}{3}, data: map[string]interface{}{"x": []interface{}{map[string]interface{}{"a": 3, "b": "sample1"}}}},
+               {path: "$.x[?(@.a==3)].b", expect: []interface{}{"sample1"}, data: map[string]interface{}{"x": []interface{}{map[string]interface{}{"a": 3, "b": "sample1"}}}},
+               // filter with struct
+               {path: "$.x[?(@.b=='sample2')].a", expect: []interface{}{3}, data: Any{X: []*Sample{&Sample{A: 3, B: "sample2"}}}},
+               {path: "$.x[?(@.a==4)].b", expect: []interface{}{"sample2"}, data: Any{X: []*Sample{&Sample{A: 4, B: "sample2"}}}},
        }
 )

```

**Comment by ohler55** [comment]
Reformtted a bit but the same code with one exception. An instead of `[]*Sample` `[]interface{}` is used. Clearly it would be better to support the former so it is a bug and will be fixed. I did see some odd behaviour when I first tried it but I think I had some odd invisible character when I first copied and pasted. The lines below worked and are in the bug/jsonpath-reflect-arrays branch.

``` golang
		// filter with map
		{
			path:   "$.x[?(@.b=='sample1')].a",
			expect: []interface{}{3},
			data:   map[string]interface{}{"x": []interface{}{map[string]interface{}{"a": 3, "b": "sample1"}}},
		},
		{
			path:   "$.x[?(@.a==3)].b",
			expect: []interface{}{"sample1"},
			data:   map[string]interface{}{"x": []interface{}{map[string]interface{}{"a": 3, "b": "sample1"}}},
		},
		// filter with struct
		{
			path:   "$.x[?(@.b=='sample2')].a",
			expect: []interface{}{3},
			data:   Any{X: []interface{}{&Sample{A: 3, B: "sample2"}}},
		},
		{
			path:   "$.x[?(@.a==4)].b",
			expect: []interface{}{"sample2"},
			data:   Any{X: []interface{}{&Sample{A: 4, B: "sample2"}}},
		},
```

**Reply by obloquy** [reply]
Thank you. I've run into similar issues in my own code and it just reinforces my dislike of Go's reflection. I was hoping to search through a quite complex concrete object, so I'll work around it by converting it to a map first, but it's a relatively expensive operation. Other than that this is the most complete and consistent jsonpath implementation I've found. A very impressive piece of work. Thanks again.

**Comment by ohler55** [comment]
Thanks for the praise. Feedback and use in various ways it the best way for me to make it better.

**Comment by ohler55** [comment]
Give branch bug/jsonpath-reflect-arrays a try.

**Reply by obloquy** [reply]
Perfect, thanks. I'm very pleased to delete my workaround code.

**Reply by ohler55** [reply]
Released v1.12.1 today with the fix. Thanks for finding that.

**Reply by obloquy** [reply]
You're welcome, and thanks for the quick turnaround.

**Comment by ngjaying** [comment]
Is it possible to get the index after filter? For example, "$.x[?(@.b=='sample2')].index" where "index" is an implicit field for the array index. If it is not supported, do you think it is a good idea to add such syntax? Thank you.

**Reply by ohler55** [reply]
I assume that you would like to get the matching path or paths to the match or matches. While that is probably useful, using a trailing `.index` is probably not the right way to get that as it would mask fields with the same name. As an alternative adding a set of functions such as `GetPaths()` or `FirstPath()` would go along well with the current functions on the `jp.Expr` type. Maybe that would be of interest.

**Reply by ngjaying** [reply]
Thank you for the quick reply. I agree it makes more sense to add functions for the feature.

**Comment by stubents** [comment]
Hi, I just noticed that a json path with open intervals and negativ steps mostly produces rather unintuitive results.
E.g. a test like
`{path: "$[::-1]", expect: []any{5, 4, 3, 2, 1}, data: []any{1, 2, 3, 4, 5}},`
looks rather reasonable to me but jp just produces an empty slice.
Admittedly, there is no overwhelming consensus, but I just wondered if this is on purpose :)

**Reply by ohler55** [reply]
That is expected. The defaults for the start and end are 0 and the length of the array. So starting at 0 with a negative step drops off the front of the array on the first try.
